src/jvmMain/kotlin/software/momento/kotlin/sdk/internal/InternalTopicClient.jvm.kt

```diff
@@ -22,30 +36,171 @@ internal actual class InternalTopicClient actual constructor(
+        var retry = true
+
         val metadata = metadataWithCache(cacheName)
+        while (retry && currentCoroutineContext().isActive) {
+            try {
+                val request = buildSubscriptionRequest(cacheName, topicName, lastSequenceNumber)
+                val subscriptionFlow = stubsManager.streamingStub.subscribe(request, metadata)
+                subscriptionFlow.onEach { message ->
+                    if (isFirstMessage) {
+                        isFirstMessage = false
+                        if (message.kindCase != _SubscriptionItem.KindCase.HEARTBEAT) {
+                            throw InternalServerException(
+                                "Expected heartbeat message for topic $topicName on cache $cacheName. Got: ${message.kindCase}"
+                            )
+                        }
+                    }
+                }.mapNotNull { message ->
+                    convertSubscriptionItem(message)
+                }.collect { (topicMessage, sequenceNumber) ->
+                    if (!isFirstMessage) {
+                        topicMessage?.let { emit(it) }
+                        sequenceNumber?.let { lastSequenceNumber = it }
+                    }
+                }
+            } catch (e: Exception) {
+                if (e is CancellationException) {
+                    throw e // Propagate cancellation
+                }
+                val sdkException = CacheServiceExceptionMapper.convert(e, metadata)
+                if (isFirstMessage) {
+                    throw sdkException
+                } else {
+                    if (isRecoverableError(sdkException)) {
+                        delay(5000)
+                    } else {
+                        emit(TopicMessage.Error(sdkException))
+                        retry = false
+                    }
+                }
+            }
+        }
+    }
 
-        val request = _PublishRequest.newBuilder()
-            .setCacheName(cacheName)
-            .setTopic(topicName)
-            .setValue(_TopicValue.newBuilder().setText(value).build())
-            .build()
+    private fun buildSubscriptionRequest(
+        cacheName: String, topicName: String, lastSequenceNumber: Long?
+    ): _SubscriptionRequest {
+        return _SubscriptionRequest.newBuilder().apply {
+            this.cacheName = cacheName
+            this.topic = topicName
+            if (lastSequenceNumber != null) {
+                this.resumeAtTopicSequenceNumber = lastSequenceNumber as Long
+            }
+        }.build()
+    }
+
+    private data class ConvertedMessage(val message: TopicMessage?, val lastSequenceNumber: Long?)
 
-        try {
-            stubsManager.stub.publish(request, metadata)
-        } catch (e: Exception) {
-            return TopicPublishResponse.Error(CacheServiceExceptionMapper.convert(e, metadata))
+    private fun convertSubscriptionItem(subscriptionItem: _SubscriptionItem): ConvertedMessage? {
```

Comment:
Just thinking out loud, would it be possible to put this method in the `InternalClient`, and just have the separate implementations of `InternalClient` use it? Would just prevent us from having to write the methods twice.